File: Java/src/test/java/com/pmangplus/test/PmangPlusTest.java
package com.pmangplus.test;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import com.pmangplus.test.result.*;
import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.message.BasicNameValuePair;
import org.junit.*;

import org.apache.http.client.HttpClient;
import org.junit.Assert;
import org.junit.Test;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;

public class PmangPlusTest {

    // application identifier
    private static final String APP_ID = "pmang app id";
    private static final String APP_KEY = "pmang app key";
    private static final String APP_SECRET = "pmang app secret";
    private static final String LOCAL_CD = "KOR";
    private static final String DEVICE_CD = "ANDROID";
    //private static final String DEVICE_CD = "IPHONE";

    // user identifier
    private static final String ID = "pmang id";
    private static final String PW = "pmang pw";

    // payment identifier
    private static final String DEVELOPER_PAYLOAD = "payment developer payload";
    private static final String PURCHASE_TOKEN = "payment purchase token";

    // QA/TQ/Live URL
    private static final String URL = "http://qa.pmangplus.com";
    //private static final String URL = "http://tq.pmangplus.com";
    //private static final String URL = "http://pmangplus.com";

    private HttpClient client = HttpClientBuilder.create().build();

    private String access_token;
    private String member_id;
    private String nickname;

    @Before
    public void setUp() throws Exception {
        // Arrange
        List<NameValuePair> params = new ArrayList<NameValuePair>(Arrays.asList(
            new BasicNameValuePair("local_cd", LOCAL_CD),
            new BasicNameValuePair("app_id", APP_ID),
            new BasicNameValuePair("app_key", APP_KEY),
            new BasicNameValuePair("app_secret", APP_SECRET),
            new BasicNameValuePair("udid", "TestUDID_Java"),
            new BasicNameValuePair("email", ID),
            new BasicNameValuePair("passwd", PW),
            new BasicNameValuePair("device_cd", DEVICE_CD)
        ));

        // Act
        LoginResult result = post("/accounts/2/login", params, LoginResult.class);

        // Assert
        Assert.assertNotNull(result.value.access_token);
        Assert.assertNotNull(result.value.member.member_id);
        Assert.assertNotNull(result.value.member.nickname);

        access_token = result.value.access_token;
        member_id = result.value.member.member_id;
        nickname = result.value.member.nickname;
    }

    @Test
    public void testMembersSelf() throws Exception {
        // Arrange
        List<NameValuePair> params = new ArrayList<NameValuePair>(Arrays.asList(
            new BasicNameValuePair("local_cd", LOCAL_CD),
            new BasicNameValuePair("access_token", access_token)
        ));

        // Act
        MembersSelfResult result = post("/members/@self", params, MembersSelfResult.class);

        // Assert
        Assert.assertEquals(member_id, result.value.member_id);
        Assert.assertEquals(nickname, result.value.nickname);
        Assert.assertNotNull(result.value.crt_dt);
        Assert.assertNotNull(result.value.upd_dt);
        Assert.assertNotNull(result.value.status_cd);
        Assert.assertNotNull(result.value.profile_img_url);
        //Assert.assertNotNull(result.value.profile_img_url_raw);
        //Assert.assertNotNull(result.value.feeling);
        Assert.assertNotNull(result.value.adult_auth_yn);
        //Assert.assertNotNull(result.value.adult_auth_dt);
        Assert.assertNotNull(result.value.recent_app_id);
        //Assert.assertNotNull(result.value.recent_app_title);
        Assert.assertNotNull(result.value.recent_login_dt);
        Assert.assertNotNull(result.value.email);
        //Assert.assertNotNull(result.value.anonymouns_yn);
        //Assert.assertNotNull(result.value.reg_path);
        //Assert.assertNotNull(result.value.last_msg_dt);
        //Assert.assertNotNull(result.value.new_msg_dt);
        Assert.assertNotNull(result.value.friend_accept_cd);
        //Assert.assertNotNull(result.value.achieve_detail_info_list);
        //Assert.assertNotNull(result.value.member_achievement_summary);
        //Assert.assertNotNull(result.value.summray);
        //Assert.assertNotNull(result.value.options);
        //Assert.assertNotNull(result.value.contact);
        Assert.assertNotNull(result.value.contact_friend_mapping);
        Assert.assertNotNull(result.value.contact_require_import);
        Assert.assertNotNull(result.value.contact_require_phone_auth);
        Assert.assertNotNull(result.value.contact_require_phone_auth_confirm);
        Assert.assertNotNull(result.value.pmang_usn);
        //Assert.assertNotNull(result.value.ci);
        Assert.assertNotNull(result.value.sanction);
    }

    @Test
    public void testMarketEffectuate() throws Exception {
        // Arrange
        List<NameValuePair> params = new ArrayList<NameValuePair>(Arrays.asList(
            new BasicNameValuePair("local_cd", LOCAL_CD),
            new BasicNameValuePair("developerPayload", DEVELOPER_PAYLOAD),
            new BasicNameValuePair("purchaseToken", PURCHASE_TOKEN)
        ));

        // Act
        MarketEffectuateResult result = post("/market/v3/effectuate", params, MarketEffectuateResult.class);

        // Assert
        // if success
        //Assert.assertEquals(result.value.result, "SUCCESS");
        //Assert.assertNotNull(result.value.store_product_id);
        //Assert.assertNotNull(result.value.pay_id);

        // if not success (FAILURE / DUPLICATED / NOTEXISTS)
        Assert.assertNotNull(result.value.result);
    }

    @Test
    public void testGcmPublish() throws Exception {
        // Arrange
        String member_ids = member_id.toString() + "|" +
                            member_id.toString() + "|" +
                            member_id.toString();

        List<NameValuePair> params = new ArrayList<NameValuePair>(Arrays.asList(
                new BasicNameValuePair("local_cd", LOCAL_CD),
                new BasicNameValuePair("member_ids", member_ids),
                new BasicNameValuePair("message", "")
        ));

        // Act
        PushResult result = post("/gcm/" + APP_ID + "/publish", params, PushResult.class);

        // Assert
        Assert.assertNotNull(result.value);
    }

    @Test
    public void testGcmGetAllow() throws Exception {
        // Arrange
        List<NameValuePair> params = new ArrayList<NameValuePair>(Arrays.asList(
                new BasicNameValuePair("local_cd", LOCAL_CD),
                new BasicNameValuePair("access_token", access_token)
        ));

        // Act
        PushResult result = post("/gcm/" + APP_ID + "/getAllow", params, PushResult.class);

        // Assert
        Assert.assertNotNull(result.value);
    }

    @Test
    public void testGcmSetAllow() throws Exception {
        Boolean push_allow = false;
        PushResult result;

        // get allow
        List<NameValuePair> getAllowParams = new ArrayList<NameValuePair>(Arrays.asList(
                new BasicNameValuePair("local_cd", LOCAL_CD),
                new BasicNameValuePair("access_token", access_token)
        ));

        result = post("/gcm/" + APP_ID + "/getAllow", getAllowParams, PushResult.class);

        Assert.assertNotNull(result.value);
        push_allow = Boolean.valueOf(result.value);

        // set allow
        List<NameValuePair> setAllowParams = new ArrayList<NameValuePair>(Arrays.asList(
                new BasicNameValuePair("local_cd", LOCAL_CD),
                new BasicNameValuePair("access_token", access_token),
                new BasicNameValuePair("allow", String.valueOf(!push_allow))
        ));

        result = post("/gcm/" + APP_ID + "/setAllow", setAllowParams, PushResult.class);

        Assert.assertNotNull(result.value);
        Assert.assertEquals(push_allow, Boolean.valueOf(result.value));

        // get allow
        result = post("/gcm/" + APP_ID + "/getAllow", getAllowParams, PushResult.class);

        Assert.assertNotNull(result.value);
        Assert.assertEquals(push_allow, Boolean.valueOf(result.value));
    }

    @Test
    public void testApnsPublish() throws Exception {
        // Arrange
        String member_ids = member_id.toString() + "|" +
                member_id.toString() + "|" +
                member_id.toString();

        List<NameValuePair> params = new ArrayList<NameValuePair>(Arrays.asList(
                new BasicNameValuePair("local_cd", LOCAL_CD),
                new BasicNameValuePair("member_ids", member_ids),
                new BasicNameValuePair("message", "")
        ));

        // Act
        PushResult result = post("/apns/" + APP_ID + "/publish", params, PushResult.class);

        // Assert
        Assert.assertNotNull(result.value);
    }

    @Test
    public void testApnsGetAllow() throws Exception {
        // Arrange
        List<NameValuePair> params = new ArrayList<NameValuePair>(Arrays.asList(
                new BasicNameValuePair("local_cd", LOCAL_CD),
                new BasicNameValuePair("access_token", access_token)
        ));

        // Act
        PushResult result = post("/apns/" + APP_ID + "/getAllow", params, PushResult.class);

        // Assert
        Assert.assertNotNull(result.value);
    }

    @Test
    public void testApnsSetAllow() throws Exception {
        Boolean push_allow = false;
        PushResult result;

        // get allow
        List<NameValuePair> getAllowParams = new ArrayList<NameValuePair>(Arrays.asList(
                new BasicNameValuePair("local_cd", LOCAL_CD),
                new BasicNameValuePair("access_token", access_token)
        ));

        result = post("/apns/" + APP_ID + "/getAllow", getAllowParams, PushResult.class);

        Assert.assertNotNull(result.value);
        push_allow = Boolean.valueOf(result.value);

        // set allow
        List<NameValuePair> setAllowParams = new ArrayList<NameValuePair>(Arrays.asList(
                new BasicNameValuePair("local_cd", LOCAL_CD),
                new BasicNameValuePair("access_token", access_token),
                new BasicNameValuePair("allow", String.valueOf(!push_allow))
        ));

        result = post("/apns/" + APP_ID + "/setAllow", setAllowParams, PushResult.class);

        Assert.assertNotNull(result.value);
        Assert.assertEquals(push_allow, Boolean.valueOf(result.value));

        // get allow
        result = post("/apns/" + APP_ID + "/getAllow", getAllowParams, PushResult.class);

        Assert.assertNotNull(result.value);
        Assert.assertEquals(push_allow, Boolean.valueOf(result.value));
    }

    private <T extends ResultBase> T post(String prefix, List<NameValuePair> params, Class<T> type) throws Exception {
        String url = URL + prefix;
        HttpPost request = new HttpPost(url);

        for (Map.Entry<String, String> header : makeHeader().entrySet()) {
            request.addHeader(header.getKey(), header.getValue());
        }

        request.setEntity(new UrlEncodedFormEntity(params));
        HttpResponse response = client.execute(request);

        Assert.assertEquals(response.getStatusLine().getStatusCode(), 200);

        BufferedReader rd = new BufferedReader(
                new InputStreamReader(response.getEntity().getContent())
        );

        StringBuffer result = new StringBuffer();
        String line = "";
        while ((line = rd.readLine()) != null) {
            result.append(line);
        }

        Gson gson = new GsonBuilder().create();
        ResultBase res = gson.fromJson(result.toString(), type);
        Assert.assertEquals(res.result_code, "000");
        Assert.assertEquals(res.result_msg, "API_OK");
        return (T)res;
    }

    private void get(String prefix, List<NameValuePair> params) {

    }

    private HashMap<String, String> makeHeader() throws NoSuchAlgorithmException {
        String now = String.valueOf(System.currentTimeMillis());
        HashMap<String, String> headers = new HashMap<String, String>();
        headers.put("ver", "4");
        headers.put("ts", now);
        headers.put("fp", sha1(now + APP_SECRET));
        headers.put("User-Agent", "PmangPlus for Server");
        return headers;
    }

    private String sha1(String input) throws NoSuchAlgorithmException {
        // Reference : http://www.sha1-online.com/sha1-java/
        MessageDigest mDigest = MessageDigest.getInstance("SHA1");
        byte[] result = mDigest.digest(input.getBytes());
        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < result.length; i++) {
            sb.append(Integer.toString((result[i] & 0xff) + 0x100, 16).substring(1));
        }

        return sb.toString();
    }
}

File: PHP/PmangPlusTest.php
<?php

require('vendor/autoload.php');

class PmangPlusTest extends PHPUnit_Framework_TestCase
{
	// application identifier
	private $APP_ID = 'pmang app id';
	private $APP_KEY = 'pmang app key';
	private $APP_SECRET = 'pmang app secret';
	private $LOCAL_CD = 'KOR';
	private $DEVICE_CD = 'ANDROID';
	//private $DEVICE_CD = 'IPHONE';

	// user identifier
	private $ID = 'pmang id';
	private $PW = 'pmang pw';

	// payment identifier
	private $DEVELOPER_PAYLOAD = 'payment developer payload';
	private $PURCHASE_TOKEN = 'payment purchase token';

	// QA/TQ/Live URL
	private $URL = 'http://qa.pmangplus.com';
	//private $URL = 'http://tq.pmangplus.com';
	//private $URL = 'http://pmangplus.com';

	private $access_token = '';
	private $member_id = '';
	private $nickname = '';


	protected function setUp()
	{
		$response = $this->post('/accounts/2/login', [
			'app_id' => $this->APP_ID,
			'app_key' => $this->APP_KEY,
			'app_secret' => $this->APP_SECRET,
			'udid' => 'TestUDID_PHP',
			'email' => $this->ID,
			'passwd' => $this->PW,
			'local_cd' => $this->LOCAL_CD,
			'device_cd' => $this->DEVICE_CD
		]);

		$value = $response->value;
		$this->access_token = $value->access_token;

		$this->assertNotNull($value->member);

		$member = $value->member;
		$this->member_id = $member->member_id;
		$this->nickname = $member->nickname;
	}

	public function testMembersSelf()
	{
		// Arrange
		// Act
		$response = $this->post('/members/@self', [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token,
		]);

		// Assert
		$value = $response->value;
		
		$this->assertEquals($value->member_id, $this->member_id);
		$this->assertEquals($value->nickname, $this->nickname);
		$this->assertTrue(property_exists($value, 'crt_dt'));
		$this->assertTrue(property_exists($value, 'upd_dt'));
		$this->assertTrue(property_exists($value, 'status_cd'));
		$this->assertTrue(property_exists($value, 'profile_img_url'));
		$this->assertTrue(property_exists($value, 'feeling'));
		$this->assertTrue(property_exists($value, 'adult_auth_yn'));
		$this->assertTrue(property_exists($value, 'adult_auth_dt'));
		$this->assertTrue(property_exists($value, 'recent_login_dt'));
		$this->assertTrue(property_exists($value, 'recent_app_id'));
		$this->assertTrue(property_exists($value, 'email'));
		$this->assertTrue(property_exists($value, 'anonymous_yn'));
		$this->assertTrue(property_exists($value, 'reg_path'));
		$this->assertTrue(property_exists($value, 'recent_app_title'));
		$this->assertTrue(property_exists($value, 'last_msg_dt'));
		$this->assertTrue(property_exists($value, 'new_msg_yn'));
		$this->assertTrue(property_exists($value, 'friend_accept_cd'));
		$this->assertTrue(property_exists($value, 'achieve_detail_info_list'));
		$this->assertTrue(property_exists($value, 'member_achievement_summary'));
		$this->assertTrue(property_exists($value, 'summary'));
		$this->assertTrue(property_exists($value, 'options'));
		$this->assertTrue(property_exists($value, 'contact_require_phone_auth'));
		$this->assertTrue(property_exists($value, 'contact_require_import'));
		$this->assertTrue(property_exists($value, 'contact_friend_mapping'));
		$this->assertTrue(property_exists($value, 'contact_require_phone_auth_confirm'));
		$this->assertTrue(property_exists($value, 'contact'));
		$this->assertTrue(property_exists($value, 'pmang_usn'));
		$this->assertTrue(property_exists($value, 'ci'));
		$this->assertTrue(property_exists($value, 'sanction'));
		$this->assertTrue(property_exists($value, 'profile_img_url_raw'));
	}
	
	public function testMarketEffectuate()
	{
		// Arrange
		// Act
		$response = $this->post('/market/v3/effectuate', [
			'local_cd' => $this->LOCAL_CD,
			'developerPayload' => $this->DEVELOPER_PAYLOAD,
			'purchaseToken' => $this->PURCHASE_TOKEN
		]);

		// Assert
		$value = $response->value;

		// if success
		//$this->assertEquals($value->result, 'SUCCESS');
		//$this->assertTrue(property_exists($value, 'store_product_id'));
		//$this->assertTrue(property_exists($value, 'pay_id'));
		
		// if not success(FAILURE / DUPLICATED / NOTEXISTS)
		$this->assertTrue(property_exists($value, 'result'));
	}

	public function testGcmPublish()
	{
		// Arrange
		$member_ids = array($this->member_id, $this->member_id, $this->member_id);

		// Act
		$response = $this->post('/gcm/' . $this->APP_ID . '/publish', [
			'local_cd' => $this->LOCAL_CD,
			'member_ids' => join("|", $member_ids),
			'message' => '{"content":"내용", "title":"제목", "type":"ad", "url":"http://ppl.cm/abcdef"}'
		]);

		// Assert
		$this->assertNotNull($response->{'value'});
	}

	public function testGcmGetAllow()
	{
		// Arrange
		// Act
		$response = $this->post('/gcm/' . $this->APP_ID . '/getAllow', [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token
		]);

		// Assert
		$this->assertNotNull($response->{'value'});
	}

	public function testGcmSetAllow()
	{
		$push_allow = false;

		// get allow
		$response = $this->post('/gcm/' . $this->APP_ID . '/getAllow', [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token
		]);

		$this->assertNotNull($response->value);
		$push_allow = $response->value;

		// set allow
		$response = $this->post('/gcm/' . $this->APP_ID . '/setAllow', [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token,
			'allow' => !$push_allow
		]);

		$this->assertNotNull($response->value);

		// get allow
		$response = $this->post('/gcm/' . $this->APP_ID . '/getAllow', [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token
		]);

		$this->assertNotNull($response->value);
		$this->assertEquals($push_allow, $response->value);
	}

	public function testApnsPublish()
	{
		// Arrange
		$prefix = '/apns/' . $this->APP_ID . '/publish';
		$member_ids = array($this->member_id, $this->member_id, $this->member_id);

		// Act
		$response = $this->post($prefix, [
			'local_cd' => $this->LOCAL_CD,
			'member_ids' => join("|", $member_ids),
			'message' => '내용'
		]);

		// Assert
		$this->assertNotNull($response->{'value'});
	}

	public function testApnsGetAllow()
	{
		// Arrange
		// Act
		$response = $this->post("/apns/" . $this->APP_ID . "/getAllow", [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token
		]);

		// Assert
		$this->assertNotNull($response->{'value'});
	}

	public function testApnsSetAllow()
	{
		$push_allow = false;

		// get allow
		$response = $this->post('/apns/' . $this->APP_ID . '/getAllow', [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token
		]);

		$this->assertNotNull($response->value);
		$push_allow = $response->value;

		// set allow
		$response = $this->post('/apns/' . $this->APP_ID . '/setAllow', [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token,
			'allow' => !$push_allow
		]);

		$this->assertNotNull($response->value);

		// get allow
		$response = $this->post('/apns/' . $this->APP_ID . '/getAllow', [
			'local_cd' => $this->LOCAL_CD,
			'access_token' => $this->access_token
		]);

		$this->assertNotNull($response->value);
		$this->assertEquals($push_allow, $response->value);
	}

	private function post($prefix, $params)
	{
		$client = new GuzzleHttp\Client([
      'base_uri' => $this->URL
		]);

		$result = $client->post($prefix, [
			'headers' => $this->makeHeader(),
			'form_params' => $params
		]);

		$response = json_decode($result->getBody());
		$this->assertEquals($response->{'result_code'}, '000');
		$this->assertEquals($response->{'result_msg'}, 'API_OK');
		$this->assertNotNull($response->{'value'});

		return $response;
	}

	private function get($prefix, $params)
	{
		// TODO: get request
	}
	
	private function makeHeader()
	{
		$now = $this->currentMillis();
		$headers = array(
			'ver' => 4,
			'ts' => $now,
			'fp' => sha1($now . $this->APP_SECRET),
			'User-Agent' => 'PmangPlus for Server'
		);

		return $headers;
	}

	private function currentMillis()
	{
		$mt = explode(' ', microtime());
		return $mt[1] * 1000 + round($mt[0] * 1000);
	}
}

?>
